### DISCUSSION TITLE
Change warn to debug when parsing non-string array claim element in oidc_util_set_app_infos

### DISCUSSION THREAD

**Discussion by nclarkau** [OpenIDC/mod_auth_openidc#1263] [General]
We have a claim (roles) in the ID token that is an array of objects and this is triggering a warning on every request as they cannot be parsed. 

    "roles": [
        {
            "type": "xxxxxxxxxxxxxxx",
            "value": "xxxxxxxxxxxxxxx"
        },
        {
            "type": "xxxxxxxxxxxxxxx",
            "value": "xxxxxxxxxxxxxxx"
        }
    ],

This means we get a tsunami of warnings since all sessions/tokens contain the same claim structure. If we try to reduce the log levels for this module we miss out on other much more useful warnings.

    [Fri Jul 19 06:52:57.034856 2024] [auth_openidc:warn] [pid 1660885:tid 140259236984576] [client 10.42.193.122:46422] oidc_util_set_app_infos: unhandled in-array JSON object type [0] for key "roles" when parsing claims array elements, referer: xxxx
    [Fri Jul 19 06:52:57.034853 2024] [auth_openidc:warn] [pid 1660885:tid 140259236984576] [client 10.42.193.122:46422] oidc_util_set_app_infos: unhandled in-array JSON object type [0] for key "roles" when parsing claims array elements, referer: xxxx
    [Fri Jul 19 06:52:57.034698 2024] [auth_openidc:warn] [pid 1660885:tid 140259236984576] [client 10.42.193.122:46422] oidc_util_set_app_infos: unhandled in-array JSON object type [0] for key "roles" when parsing claims array elements, referer: xxxx
    [Fri Jul 19 06:52:57.034695 2024] [auth_openidc:warn] [pid 1660885:tid 140259236984576] [client 10.42.193.122:46422] oidc_util_set_app_infos: unhandled in-array JSON object type [0] for key "roles" when parsing claims array elements, referer: xxxx
    [Fri Jul 19 06:52:57.003063 2024] [auth_openidc:warn] [pid 1660886:tid 140259270555392] [client 10.42.193.122:12408] oidc_util_set_app_infos: unhandled in-array JSON object type [0] for key "roles" when parsing claims array elements, referer: xxxx
    [Fri Jul 19 06:52:57.003060 2024] [auth_openidc:warn] [pid 1660886:tid 140259270555392] [client 10.42.193.122:12408] oidc_util_set_app_infos: unhandled in-array JSON object type [0] for key "roles" when parsing claims array elements, referer: xxxx
    [Fri Jul 19 06:52:57.002909 2024] [auth_openidc:warn] [pid 1660886:tid 140259270555392] [client 10.42.193.122:12408] oidc_util_set_app_infos: unhandled in-array JSON object type [0] for key "roles" when parsing claims array elements, referer: xxxx
    [Fri Jul 19 06:52:57.002905 2024] [auth_openidc:warn] [pid 1660886:tid 140259270555392] [client 10.42.193.122:12408] oidc_util_set_app_infos: unhandled in-array JSON object type [0] for key "roles" when parsing claims array elements, referer: xxxx
    [Fri Jul 19 06:52:56.778863 2024] [auth_openidc:warn] [pid 1660885:tid 140259245377280] [client 10.42.193.53:34374] oidc_util_set_app_infos: unhandled in-array JSON object type [0] for key "roles" when parsing claims array elements, referer: xxxx
    [Fri Jul 19 06:52:56.778863 2024] [auth_openidc:warn] [pid 1660885:tid 140259245377280] [client 10.42.193.53:34374] oidc_util_set_app_infos: unhandled in-array JSON object type [0] for key "roles" when parsing claims array elements, referer: xxxx

It kind of goes against principles of log level verbosity where warnings should be infrequent and not on every transaction/request. As I suspect in most cases, just like we have found, if this condition is met its highly likely to be met on all requests.

code reference:
https://github.com/OpenIDC/mod_auth_openidc/blob/94f832f5ab6dbacec88094ea12931ed7fdc49f8e/src/util.c#L1625

**Could this please be changed to debug instead of warn?**

**Comment by zandbelt** [comment]
How about blacklisting the claim with [`OIDCBlackListedClaims` ](https://github.com/OpenIDC/mod_auth_openidc/blob/v2.4.16.3/auth_openidc.conf#L1077-L1080) or even better, have the IDP not issue unused claims at all?

**Comment by nclarkau** [comment]
The claim is integral to the identity solution so we need it in ID tokens sent downstream. It looks like **OIDCBlackListedClaims** would also remove from the ID token.

**Reply by zandbelt** [reply]
I see, it is now changed in ea3af872dcdbb4634a7e541c5e8c7326dafbb090

**Reply by nclarkau** [reply]
Thank you!

**Reply by zandbelt** [reply]
this is now released in https://github.com/OpenIDC/mod_auth_openidc/releases/tag/v2.4.16.4

**Comment by patricksouth** [comment]
I have also seen this same error with nested claims, that are valid json.
I'm curious why mod_auth_openidc won't unpack [nclarkau](https://github.com/nclarkau)'s claim (roles) ?
does it not conform to the OIDC spec?
thanks

**Reply by zandbelt** [reply]
which version of mod_auth_openidc are you using?

**Comment by patricksouth** [comment]
currently 2.4.17

**Comment by zandbelt** [comment]
The idea behind passing claims in headers is that they can be consumed  as name/value pairs easily by applications, without requiring them to parsing JSON, that is why only (mostly) basic types are supported. If your application requires consuming a nested JSON object, you'll need to parse JSON in the application anyhow and you can use `OIDCPassUserInfoAs json`  (and/or `OIDCPassIDTokenAs payload`) to pass the full contents of the Userinfo and/or ID Token in JSON format.
